File: src/main/java/caelum/focalworks/Focalworks.java
package caelum.focalworks;

import at.petrak.hexcasting.api.casting.RenderedSpell;
import at.petrak.hexcasting.api.casting.eval.CastingEnvironment;
import at.petrak.hexcasting.api.casting.eval.vm.CastingImage;
import at.petrak.hexcasting.api.casting.eval.vm.CastingVM;
import caelum.focalworks.registry.FocalworksIoticblocksActions;
import net.fabricmc.api.ModInitializer;

import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import caelum.focalworks.registry.FocalworksActions;

import java.util.HashMap;
import java.util.List;

public class Focalworks implements ModInitializer {
    public static final ThreadLocal<HashMap<String,Object>> CONTEXT = ThreadLocal.withInitial(HashMap::new);
	public static final String MOD_ID = "focalworks";
	// This logger is used to write text to the console and the log file.
	// It is considered best practice to use your mod id as the logger's name.
	// That way, it's clear which mod wrote info, warnings, and errors.
	public static final Logger LOGGER = LoggerFactory.getLogger(MOD_ID);

    public static ResourceLocation id(String path) {
        return new ResourceLocation(MOD_ID,path);
    }

	@Override
	public void onInitialize() {
		// This code runs as soon as Minecraft is in a mod-load-ready state.
		// However, some things (like resources) may still be uninitialized.
		// Proceed with mild caution.
        FocalworksActions.init();
        if (FabricLoader.getInstance().isModLoaded("ioticblocks")) {
            FocalworksIoticblocksActions.init();
        }
		LOGGER.info("Focalworks' actions registering!");
	}
    public static int clamp(int val, int min, int max) {
        return Math.min(max,Math.max(val, min));
    }
    public static RenderedSpell emptyRenderedSpell = new RenderedSpell() {
        @Override
        public void cast(@NotNull CastingEnvironment castingEnvironment) {

        }

        @Override
        public @Nullable CastingImage cast(@NotNull CastingEnvironment castingEnvironment, @NotNull CastingImage castingImage) {
            return null;
        }
    };
}

File: common/src/main/kotlin/io/github/c_caelum/focalworks/Focalworks.kt
package io.github.c_caelum.focalworks

import at.petrak.hexcasting.api.casting.iota.IotaType
import at.petrak.hexcasting.api.casting.iota.PatternIota
import at.petrak.hexcasting.api.casting.math.HexDir
import at.petrak.hexcasting.api.casting.math.HexPattern
import net.minecraft.resources.ResourceLocation
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import io.github.c_caelum.focalworks.config.FocalworksServerConfig
import io.github.c_caelum.focalworks.networking.FocalworksNetworking
import io.github.c_caelum.focalworks.registry.FocalworksActions

object Focalworks {
    const val MOD_ID = "focalworks"

    @JvmField
    val LOGGER: Logger = LogManager.getLogger(MOD_ID)

    @JvmStatic
    fun id(path: String) = ResourceLocation(MOD_ID, path)

    fun init() {
        LOGGER.warn(IotaType.serialize(PatternIota(HexPattern.fromAngles("de", HexDir.NORTH_EAST))))
        FocalworksServerConfig.init()
        initRegistries(
            FocalworksActions,
        )
        FocalworksNetworking.init()
    }

    fun initServer() {
        FocalworksServerConfig.initServer()
    }
}
